### PR TITLE
SOAP with history

### PR DESCRIPTION
**Pull request by boxlady**
Fixes #

#### Short description of what this resolves:
This adds history to the soap. Which shows a list of all the previous soap entries for this patient above the entry textbox.

#### Changes proposed in this pull request:
Added entry history to the SOAP form

below added by bradymiller:
Up For Grabs demo for this PR is here: https://www.open-emr.org/wiki/index.php/Development_Demo#Gamma_-_Up_For_Grabs_Demo

### CODE REVIEW

File: interface/forms/soap/FormSOAP.class.php [hunk 1]
```diff
@@ -64,117 +69,190 @@ function __construct($id = "", $_prefix = "")
...
+        $this->histroy();
     }
+
     function populate()
     {
         parent::populate();
         //$this->temp_methods = parent::_load_enum("temp_locations",false);
     }
 
+    function histroy()
```
Comment (stephenwaite) [inline interface/forms/soap/FormSOAP.class.php]: hi @boxlady :), simple typo here

File: interface/forms/soap/templates/general_new.html [hunk 2]
```diff
@@ -1,68 +1,72 @@
 <html>
 <head>
-<title>{xlt t='SOAP'}</title>
-{headerTemplate}
+    <title>{xlt t='SOAP'}</title>
+    {headerTemplate}
 </head>
```
Comment (stephenwaite) [inline interface/forms/soap/templates/general_new.html]: maybe title with `SOAP history` too?

File: interface/forms/soap/FormSOAP.class.php [hunk 3]
```diff
@@ -35,15 +35,20 @@ class FormSOAP extends ORDataObject
      */
```
Comment (stephenwaite) [inline interface/forms/soap/FormSOAP.class.php]: hi @boxlady , may want to create new with `interface/forms/soap_history/FormSOAPHistory.class.php` so preserve legendary SOAP :)

**Comment by bradymiller** [comment]
hi @boxlady and @stephenwaite ,
I haven't tested this, but like the idea of having history available from prior forms. This can probably be worked in a way where a new form does not need to be created (since it is just showing prior entries). Maybe a tooltip type graphic that expands to show the history when clicked on. Will plan to test this out soon. -brady

**Comment by bradymiller** [comment]
Up For Grabs demo for this PR is here: https://www.open-emr.org/wiki/index.php/Development_Demo#Gamma_-_Up_For_Grabs_Demo

**Comment by sunsetsystems** [comment]
Nice. For what it's worth LBF forms also have the ability to show past instances; see Repeats in the layout properties.

**Comment by stephenwaite** [comment]
looks like when you edit it doubles up the saved history

**Comment by boxlady** [comment]
> looks like when you edit it doubles up the saved history

You're right, if you click edit from soap box, it shows double, however if you open  it though the menu not.. I'll have to look into that.

**Comment by boxlady** [comment]
I updated this, removed doubles, and ID's. It has been working in production for the last couple months. Is this something that can be added to master?

**Comment by bradymiller** [comment]
hi @boxlady ,
Just tested this. Many users will not use this, however, it is a really neat thing to have and am guessing a significant amount of users will find it useful. So it doesn't detract for the users that don't use it, seems like would be best to collapse the history parts of the forms, which the user can then click on to expand them.
@stu01509 , since you are turning into the js/bootstrap expert, any thoughts on how to best do this?

**Comment by stu01509** [comment]
Hello 

Thanks for your mention, How can I test and run this PR on my environment?

**Comment by stephenwaite** [comment]
hi @stu01509, for the easy dev enviro go to your local repo
0. docker-compose down -v
1. git remote add boxlady https://github.com/boxlady/openemr
2. git fetch boxlady
3. git checkout boxlady/soep_history
4. docker pull openemr/openemr:flex
5. docker-compose up -d
6. have a cup of :coffee: 
7. docker logs <openemr container id>
should show this after you've had a sip or two
Screenshot from 2020-06-05 16-34-40
8. go to http://localhost:8300/interface/login/login.php?site=default
9. create an encounter and choose a clinical->soap form

**Comment by stu01509** [comment]
Hi @boxlady  @bradymiller @stephenwaite  and @tywrenn 

I'm so sorry about that, I overlook this issue mention. Now I have deployed this version on my environment and try to use the accordion design the sample, What do you think, let me know your thought.

image

**Comment by bradymiller** [comment]
hi @stu01509 , I would think of terms of potentially 10-100 possible prior notes. So, to be least invasive, would probably just have a panel that says something like; click here to see prior <number message>, which would then expand if the user clicks on it. That way, it won't take up much space when not used since many users likely will just ignore it.

**Comment by tywrenn** [comment]
@bradymiller @stu01509 For this situation I like accordions yet with the possibility of 10-100 of these seems like a lot. For each of the accordion btn's you should add the following classes: `btn-sm p-0`

This will change the button size to small and remove the padding so the accordion appears smaller. The output looks something like this:
image

**Comment by bradymiller** [comment]
Not sure how accordion acts. should:
1. place all the prior entries under one accordion (the provider can then see them all in one glance to quickly copy what text they want to).
2. don't lose visibility of the text field for quick pasting

**Comment by stu01509** [comment]
I guess it may look like this, Only click one accordion it will show all of the history messages.

image

**Comment by bradymiller** [comment]
@stu01509 , looks great!

**Comment by stu01509** [comment]
Hi @bradymiller 

Happy to hear that, so, how do I create this feature?

**Comment by bradymiller** [comment]
bump

**Comment by bradymiller** [comment]
bump

**Pull request closed without merging**